client/src/retry.rs

```diff
@@ -201,7 +229,26 @@ where
         let long_poll_allowed =
             is_long_poll && [Code::Cancelled, Code::DeadlineExceeded].contains(&e.code());
 
-        if RETRYABLE_ERROR_CODES.contains(&e.code()) || long_poll_allowed {
+        // Sometimes we can get a GOAWAY that, for whatever reason, isn't quite properly handled
+        // by hyper or some other internal lib, and we want to retry that still. We'll retry that
+        // at most once. Ideally this bit should be removed eventually if we can repro the upstream
+        // bug and it is fixed.
+        let mut goaway_retry_allowed = false;
+        if !self.have_retried_goaway_cancel && e.code() == Code::Cancelled {
+            if let Some(e) = e
+                .source()
+                .and_then(|e| e.downcast_ref::<tonic::transport::Error>())
+                .and_then(|te| te.source())
+                .and_then(|tec| tec.downcast_ref::<hyper::Error>())
+            {
+                if format!("{e:?}").contains("connection closed") {
```

Comment:
There is no other good way to do this without again down-casting _this_ to a &'static str

Reply:
I think the bigger concern is if they change the string on us since we can't really test this